infrastructure/channels/serialize_child.html

```diff
@@ -0,0 +1,167 @@
+        }
+    } else if (expectedValue.constructor.name === "Set") {
+        seenActual = seen.has(actualValue);
+        seenExpected = seen.has(expectedValue)
+        if (seenActual && seenExpected) {
+            return;
+        } else if (seenExpected && !seenActual) {
+            throw new Error(`Expected cyclic set`);
+        } else if (!seenExpected && seenActual) {
+            throw new Error(`Got unexpected cyclic set`);
+        }
+        seen.add(actualValue);
+        seen.add(expectedValue);
+
+
+        if (actualValue.size !== expectedValue.size) {
+            throw new Error(`Set sizes differ, expected ${expectedValue.size}, got ${actualValue.size}`);
+        }
+        // So for a set we can't really check if the values are the same
+        // except where they're primitives
+        for (let entry of expectedValue) {
+            if (["undefined", "string", "boolean", "number", "bigint"].includes(typeof entry) || entry === null) {
+                if(!actualValue.has(entry)) {
+                    throw new Error(`Set missing entry, expected ${entry}`);
+                }
+            } else {
+                throw new Error(`Can't compare non-primitive value ${entry} inside sets`);
+            }
+        }
+    } else if (expectedValue.constructor.name === "Map") {
+        seenActual = seen.has(actualValue);
+        seenExpected = seen.has(expectedValue)
+        if (seenActual && seenExpected) {
+            return;
+        } else if (seenExpected && !seenActual) {
+            throw new Error(`Expected cyclic map`);
+        } else if (!seenExpected && seenActual) {
+            throw new Error(`Got unexpected cyclic map`);
+        }
+        seen.add(actualValue);
+        seen.add(expectedValue);
+
+        if (actualValue.size !== expectedValue.size) {
+            throw new Error(`Map sizes differ, expected ${expectedValue.size}, got ${actualValue.size}`);
+        }
+        // So for a set we can't really check if the values are the same
+        // except where they're primitives
+        for (let [key, value] of expectedValue.entries()) {
+            if(!actualValue.has(key)) {
+                throw new Error(`Map missing key, expected key ${key} with value ${value}`);
+            }
+            compareValue(actualValue.get(key), value, seen);
+        }
+    } else {
+        seenActual = seen.has(actualValue);
+        seenExpected = seen.has(expectedValue)
+        if (seenActual && seenExpected) {
+            return;
+        } else if (seenExpected && !seenActual) {
+            throw new Error(`Expected cyclic object`);
+        } else if (!seenExpected && seenActual) {
+            throw new Error(`Got unexpected cyclic object`);
+        }
+        seen.add(actualValue);
+        seen.add(expectedValue);
+
+
+        // Compare as a general Object
+        if (Object.keys(actualValue).length !== Object.keys(expectedValue).length) {
```

Comment:
Calling both `Object.keys(expectedValue)` here and `Object.entries(expectedValue)` is a bit wasteful. Now, if one wanted to optimize this one should probably avoid creating those temporary arrays entirely, but let's do the simple optimization at least.